### PR TITLE
[FLINK-19737][table] Introduce TableOperatorWrapperGenerator to translate transformation DAG in a multiple-input node to TableOperatorWrapper DAG

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Transformation is not serializable, while StreamOperatorFactory is. We need to introduce another class (named TableOperatorWrapper) to store the information of a Transformation, and introduce a utility class (named TableOperatorWrapper) to convert the Transformation DAG to TableOperatorWrapper DAG. This pr aims to support it.*

## Brief change log

  - *Introduce TableOperatorWrapper to store the information of a Transformation*
  - *Introduce TableOperatorWrapperGenerator to translate transformation DAG in a multiple-input node to TableOperatorWrapper DAG*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added TableOperatorWrapperTest to test TableOperatorWrapper*
  - *Added TableOperatorWrapperGeneratorTest to test TableOperatorWrapperGenerator*
  - *Extended ExplainTest to verify the logic of BatchExecMultipleInputNode#translateToPlanInternal*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

**Review by TsReaper — COMMENTED** [review]
Thanks @godfreyhe for the commit. My main concern about this is that how do we deal with transformations with different parallelisms in the same multiple input node. See my comments in `TableOperatorWrapper` for an example.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/multipleinput/TableOperatorWrapperGenerator.java [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+
+	/**
+	 * The read order corresponding to each transformation in {@link #inputTransforms}.
+	 */
+	private final int[] readOrders;
+
+	/**
+	 * Reordered input transformations which order corresponds to the order of {@link #inputSpecs}.
+	 */
+	private final List<Transformation<?>> orderedInputTransforms;
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/multipleinput/TableOperatorWrapperGenerator.java:66]: we should keep `InputSpec` clean. we can return `List<Pair<Transformation<?>, InputSpec>>` instead

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/multipleinput/TableOperatorWrapperGenerator.java [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+	}
+
+	private void calcParallelismAndResource(Transformation<?> transform) {
+		int currentParallelism = transform.getParallelism();
+		if (parallelism < 0) {
+			parallelism = currentParallelism;
+		} else {
+			checkState(
+					currentParallelism < 0 || parallelism == currentParallelism,
+					"Parallelism of a transformation in MultipleInputNode is different from others. This is a bug.");
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/multipleinput/TableOperatorWrapperGenerator.java:189]: good catch, `set the parallelism to the maximum parallelism of the members` looks good to me. but we should add more strict validation for the plan when constructing multiple input node.

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Review by TsReaper — APPROVED** [review]
Looks good to me. Let's wait for the tests to pass.

**Pull request merged**